### PR TITLE
Removed distributed training support

### PR DESCRIPTION
**Pull request by tilmankamp**
This PR also adds support for automatically checkpointing the epoch-model with best dev-loss so far.
@reuben Main activity is in the training-loop and I tried to keep changes there - apart from trivial removal of stuff (coordinator and config/FLAGS etc.). Especially the feeder is kept as it was/is. So rebasing your big data-PR is hard for the training part but should still be possible.

### CODE REVIEW

**Comment by reuben** [comment]
I won’t have time or enough internet access to review this until Monday, but could you make sure you remove all references in READMEs and docs? Maybe nix the docs/ folder altogether? It’s severely out of date at this point.

File: DeepSpeech.py [hunk 1]
```diff
@@ -388,9 +384,11 @@ def train(server=None):
     train_set = DataSet(train_data,
                         FLAGS.train_batch_size,
                         limit=FLAGS.limit_train,
-                        next_index=lambda i: coord.get_next_index('train'))
+                        next_index=train_index.inc)
```
Comment (reuben) [inline DeepSpeech.py]: nit: `next_index=lambda i: train_index += 1` and then we can avoid defining the `SampleIndex` class and just make this variable an integer.

File: DeepSpeech.py [hunk 2]
```diff
@@ -401,7 +399,7 @@ def train(server=None):
     dev_set = DataSet(dev_data,
                       FLAGS.dev_batch_size,
                       limit=FLAGS.limit_dev,
-                      next_index=lambda i: coord.get_next_index('dev'))
+                      next_index=dev_index.inc)
```
Comment (reuben) [inline DeepSpeech.py]: Same here.

File: DeepSpeech.py [hunk 3]
```diff
@@ -491,156 +428,140 @@ def end(self, session):
...
+    # Summaries
+    step_summaries_op = tf.summary.merge_all('step_summaries')
+    step_summary_writers = {
+        'train': tf.summary.FileWriter(os.path.join(FLAGS.summary_dir, 'train'), max_queue=120),
+        'dev': tf.summary.FileWriter(os.path.join(FLAGS.summary_dir, 'dev'), max_queue=120)
+    }
 
-            current_epoch = coord._epoch-1
+    # Checkpointing
+    epoch_saver = tf.train.Saver(max_to_keep=FLAGS.max_to_keep)
```
Comment (reuben) [inline DeepSpeech.py]: Previously we had checkpoints saved every 10 minutes so that we could recover from a crash without losing too much training time. This is now only saving on every epoch end. Saving checkpoints at the end of epochs is nice to have, but I think saving every 10 minutes is a good safety feature to keep.

File: DeepSpeech.py [hunk 4]
```diff
@@ -878,53 +799,17 @@ def do_single_file_inference(input_file_path):
...
+        with tf.Graph().as_default():
+            tf.set_random_seed(FLAGS.random_seed)
+            train()
+
+    if FLAGS.test:
+        with tf.Graph().as_default():
+            test()
+
+    if FLAGS.export_dir:
+        export()
```
Comment (reuben) [inline DeepSpeech.py]: This change is just lovely :D

File: DeepSpeech.py (hunk 1 above)
Comment (tilmankamp) [reply DeepSpeech.py]: This is what I first tried. In Python `x += 1` is not resulting in a number and there is no non-hacky way to have multiple instructions in lambdas. If you go for an inline function instead, you've to make `train_index` global, as in Python non-local scope variables are only mutable if they are global. You'd have to do all of this twice for train and dev and it will get more verbose than the existing (cleaner) solution. This is Python...

File: DeepSpeech.py (hunk 3 above)
Comment (tilmankamp) [reply DeepSpeech.py]: Will address this...

File: DeepSpeech.py (hunk 1 above)
Comment (reuben) [reply DeepSpeech.py]: Ah, that's unfortunate. Alright!

**Comment by reuben** [comment]
Test failures were just version checks due to missing tags in your fork, merging.
